Update schema to snabb-softwire-v2

tsyesika opened the pull request:
This PR is to update the `lwaftr` branch to include the snabb-softwire-v2 schema which has the `br-address` leaf-list removed and put on the softwire instance and the `psid-map` list folded into the softwire.

plajjan committed:
Remove mandatory statement for list keys

List keys are implicitly mandatory so no need to declare it an extra
time.

Part of #636

dpino committed:
Sanitize YANG lists

- Check key indentifiers don't occur more than once.
- Leafs that are part of the list key are implicitly mandatory.

Commit:
#683 Add version 2 of snabb-softwire yang file

Commit:
#683 Add config migration for snabb-schema-v2

Commit:
#683 Convert the lwaftr test configs to v2

Commit:
#683 Update generate_binding_table.lua for config v2

Commit:
#683 - Fix typos in generate_binding_table.lua

This fixes the problem introduced in the last commit which also introduced
a number of typos stopping compilation and running.

Commit:
#683 - Fix snabb-softwire-v2 module name & add revision

This fixes the snabb-softwire-v2 module name a long with adding
a reivison explaining the change to the schema.

Commit:
#683 - Add mandatory to br-address leaf in snabb-softwire-v2

Commit:
#683 - Convert codebase to use snabb-softwire-v2

Commit:
#683 - Generalize migration and fix for printing

The previous migration changes assumed br-address would not be mandatory
meaning that printing would fail. This also wouldn't work for migration
from anything prior to configs at 3.0.1bis. This fixes this in a general
fashion so that any pre-(snabb-softwire-v2) configs can be migrated with
the same code and be parsable with the vanila snabb-softwire-v2 parser.

Commit:
#683 - Convert config commands to snabb-softwire-v2

This converts the commands to allow for snabb-softwire-v2 as well as the
translators between other configs (namely ietf-softwire) and the v2 version.

This drops support for the older snabb-softwire-v1 config commands.

Commit:
#683 - Convert documentation and some tests to snabb-softwire-v2

Commit:
#683 - Migrate snabbvmx configs and some more tests

This migrates the configs in snabbvmx and fixes the test (again) in
for binding_table.lua

Commit:
#683 - Migrate lwaftr's vlan test configs

Commit:
#683 - Correct the version number for config migration

dpino committed:
Merge pull request #718 from dpino/remove-mandatory-list-keys

Remove mandatory statement for list keys

dpino committed:
Merge pull request #725 from dpino/lwaftr-kristian

Merge lwaftr into lwaftr-kristian

Commit:
Fixes problem with large migration files

Commit:
Remove comments and br_address list

Commit:
Add new version to migrate-configuration README

Commit:
Remove unused local variable

Commit:
Retrigger tests to run

Commit:
Merge pull request #721 from Igalia/683-remove-br-addresses

Remove br addresses leaf-list

tsyesika committed:
Merge remote-tracking branch 'origin/lwaftr' into lwaftr-psidmap-on-softwire

tsyesika committed:
Move psidmap onto softwire in snabb-softwire-v2

This moves the psidmap onto the softwire instead of being its own block.
The "pisd" leaf still exists on the softwire directly rather than the
new "port-set" container as I believe all keys must leaves of the list
(section 7.8.2 of RFC6020).

tsyesika committed:
Add migration for psidmap on softwire

This adds a minor version for 3.2.0 (3.2.0.1) which moves the psidmap
onto the softwire. This should migrate all configurations away from
using the psidmap block and remove the "end-addr" parameter for good.

The migration can be run on a file using:

    snabb lwaftr migrate-configuration -f 3.2.0.1 <config file path>

This presumes the configuration file is already at 3.2.0 (with the
br-address leaf-list removed).

tsyesika committed:
Convert lwaftr's data files to 3.2.0.1

This converts the existing conf files for the lwaftr's tests to version
3.2.0.1. This removes the psidmap block and adds it onto the new
"port-set" container on the softwire itself.

tsyesika committed:
Convert v2 migration to handle 3.2.0 and 3.2.0.1

This converts the migration added for 3.2.0.1 to be part of the 3.2.0
snabb-softwire-v2 migration. This handles both adding "br-address" on
softwire (away from leaf-list) and psid-map to softwire.

This will not convert something converted "part way" with a previous
version of the migration. If you've already converted a config away from
the br-address list but not yet to the integrated psid-map on softwire
then you'll need to checkout a prior commit to this one and use the
3.2.0.1 converter.

This also removes the bump version which was only used in development.

tsyesika committed:
Update revision statement to include new port-set

This updates the revision statement to bump the date to when the latest
change occured. It also includes the information about the switch from
using the "psid-map" list over to using the "port-set" container on the
softwire itself. This has been done in the same revision as the
br-address as there has been no release with these changes inplace.

tsyesika committed:
Support new softwire in support and binding table

This adds support in the snabb-softwire-v2 support for the new format
where it formats new softwires as a blob the binding table can
quickly add. It also adds a transformer for initial loads of a config
to the binding table so it's able to consume configs in the new
format.

NB: This does not yet handle adding a softwire with a PSID which is
not yet known to data plane (i.e. does not appear on another
softwire).

tsyesika committed:
Change binding table's psid map to ctable

The new snabb-softwire-v2 config no longer allows for ipv4 address
ranges. This means there is no real reason to continue using the
rangemap system and it can be switched to a plain ctable.

This will give us the ability to add port-set entries on the fly
without restarting the lwaftr (how it was previously done) or
rebuilding the entire rangemap (impractical).

tsyesika committed:
Add ability to add psid entries and update tests

This adds the ability to add and remove psidmaps without restarting the
lwaftr. This allows for the addition of softwires with a new psidmapping
without requiring any additional restart. It also updates the unit tests
for the binding table to reflect the new configuration and some of the
internal changes to it.

There still remains quite a bit to do, this does NOT address:

- The leader's version of the binding table isn't updated
- No longer used versions of the PSID map are not removed
- Validation doesn't occur to check the port-set entries line up with
other port-set entries (what should even happen if they don't?)
- Translation layer isn't updated
- Some tests are failing which needs addressing.

tsyesika committed:
Updates the test for br-address on softwire

Commit:
Merge pull request #794 from Igalia/lwaftr-update-kristian

Update lwaftr-kristian with lwaftr branch

tsyesika committed:
Remove validate_config and validate_update

These might need to come back however right now they only validated that
the softwire that was being added had a corrisponding psid-map. That no
longer applies.

tsyesika committed:
Fix migration to 3.2.0

The migration to convert files pre-snabb-softwire-v2 (3.2.0) has broken
when i changed the binding_table.lua:load function. This now takes an
older version of it and uses that in order to build the binding table so
it can migrate the configurations.

tsyesika committed:
Update snabbvmx tests to the psid-map on softwires

tsyesika committed:
Fix a test to use snabb-softwire-v2 schema

This fixes an `snabb config add` test to use the updated schema. This
means adding the psid-map on the softwire using the new `port-set`
container. Previously it was erroring as it was using the old config.

This also removes the no longer valid test to check adding a softwire
without a `psid-map` entry was invalid as entires are now on the
softwire. Whilst this reduces the number of tests I think there will be
additional tests in the (near) future when the behavour around adding
softwires with different port-set info to another softwire with the same
address is decided.

tsyesika committed:
Fix a few more tests and remove redundent tests

This removes some of the tests to check if softwires without psidmaps
errored as it doesn't make as much sense with the change. It also fixes
a few other little issues with some config tests.

Outstanding: ietf-softwire translation layer is yet to do so those tests
rightly fail.

tsyesika committed:
Fix small nit in config test

Commit:
Merge pull request #796 from Igalia/lwaftr-kristian-update

Fix small nit in test on lwaftr-kristian

tsyesika committed:
Fix config translation layer for ietf-softwire

This updates the ietf-softwire translation layer for the
snabb-softwire-v2 configuration now using the `port-set` container
on `softwire` in place of the `psid-map` list.

Commit:
Fix small nits

Commit:
Update documentation for new schema

Commit:
Fix typo nit in docs

Commit:
Remove shift and padding parameters

Commit:
Update test data for removal of shift and padding

Commit:
Update migration for removal of shift and padding

Commit:
Update snabbvmx configs to remove shift too

Commit:
Fix migration to handle larger files and add padding leaf

This fixes the migration so it uses less memory, allowing for larger
files to be migrated. It's currently been tested with a 100MB file.

This also adds back the padding leaf as it causes issues in the ctable
streaming lookup of the softwires. The entry must confirm to certain
sizes or multi_copy.dasl used by ctable streaming will assert an error.

Commit:
Convert the binding table to use rangemap

This converts the binding table to use rangemap and reintroduces the
padding parameter to snabb-softwire-v2. The rangemap was reintroduced
over the ctable due to the performance impact, as rangemap uses binary
search to find the port-sets, it's notably faster.

The padding parameter was re-introduced as multi_copy.dasl complains
if the size of a ctable (in this case softwire) is not divisable by 4
when multiplied by the entries_per_lookup (max_displacement + 1).

Commit:
Fix small documentation and other nits

This fixes some of the documentation surrounding the configuration
change. This includes both user facing documentation as well as
commends throughout the code.

It also removes the newly introduced "ignore_duplicates" flag in the
RangeMapBuilder in favour of building a ctable of the added entries in
order to verify that the softwires with the same port-set definition
don't get added multiple times to the psid-map and that they don't
define different values for the port-set.

It removes two now unused functions which were to add or remove psid
map entries. These were removed as the switch back to rangemap makes
it unfeasible to modify the psid-map in the binding table once it's
been built.

wingo committed:
Fix structural typing of YANG data

YANG data should be structurally typed.  If I have a struct with just a
"uint16_t a" member, then any other struct that has that same shape
should have the same FFI type.  To implement this we have to memoize the
string -> type function so that all types with the same string
representation use the same type object.  However this is a bit tricky,
givent that types can contain members with compound types -- before,
these nested members with compound types were being treated as nominal
fresh types instead of structural types.  This commit fixes that bug.

Commit:
Add test for lwaftr's generate-binding-table subcommand

This adds a test which runs the generate-binding-table command and
verifies that it exits with a 0 status code and produces the expected
output for the command line parameters.

It doesn't check the contents of the block but it does check certain
expected things which should verify some problems if they occur.

wingo committed:
Merge pull request #807 from Igalia/cltype

Fix structural typing of YANG data

Commit:
Fix small nit in generate_binding_table_test.py

There were a few small nits found such as not including the \n when
checking if something ends in a curly brace and an undefined variable
instead of checking the length of the list.

Commit:
Update generate-binding-table for new schema changes

This updates lwaftr's generate-binding-table subcommand to include the
psid-map in the softwire to reflect the changes made in the
snabb-softwire-v2 schema changes.

Commit:
Fix small performance nit

This fixes a small performance nit in binding_table.lua which would
have been caused by allocating the key of the psid-map on every
lookup. This now reverts it back to how it was, creating the key only
when the binding table is initiated.

This should improve the performance of the BindingTable:lookup
interface a little.

Commit:
Update migrations fixing the port-set definition.

Commit:
Fix lwaftr's migrate-configuration.

This fixes lwaftr's mgirate-configuration commands going from older
versions to more recent. There have been some changes throughout the
code which means it hasn't worked but it should now.

It's been tested with:
https://people.igalia.com/dpino/lwaftr/lwaftr.conf

Commit:
Fix lwaftr's test vlan configs

Commit:
Merge pull request #795 from Igalia/lwaftr-psidmap-on-softwire

Move the psid-map to softwire in snabb-softwire-v2

wingo commented on src/apps/lwaftr/binding_table.lua:
```diff
@@ -113,7 +106,7 @@ function BTLookupQueue:process_queue()
          local ipv4 = streamer.entries[n].key.ipv4
          local port = streamer.entries[n].key.psid
-         streamer.entries[n].key.psid = self.binding_table:lookup_psid(ipv4, port)
+         streamer.entries[n].key.psid = self.binding_table:lookup_psid(ipv4, port) or 0
```
is this an intentional change?

wingo commented on src/apps/lwaftr/binding_table.lua:
```diff
@@ -163,9 +156,9 @@ function BindingTable:remove_softwire_entry(entry_key_blob)
+
 function BindingTable:lookup(ipv4, port)
-   local psid = self:lookup_psid(ipv4, port)
+   local psid = self:lookup_psid(ipv4, port) or 0
```
likewise

wingo commented on src/apps/lwaftr/binding_table.lua:
```diff
@@ -197,14 +190,10 @@ function BindingTable:lookup_psid(ipv4, port)
 -- table. Invoke like:
 --
---   for ipv4_lo, ipv4_hi, psid_info in bt:iterate_psid_map() do ... end
+--   for entry in bt:iterate_psid_map() do ... end
```
afaiu this change needs to be reverted

wingo commented on src/apps/lwaftr/rangemap.lua:
```diff
@@ -104,7 +104,7 @@ function RangeMapBuilder:build(default_value)
       local entry = self.entries[i]
       if entry.max.key >= range_end.key then
-         error("Multiple range map entries for key: "..entry.max.key)
+	 error("Multiple range map entries for key: "..entry.max.key)
```
this is an extraneous tab i think

wingo commented on src/lib/yang/data.lua:
```diff
@@ -349,7 +434,7 @@ local function table_parser(keyword, keys, values, string_key, key_ctype,
       for k,_ in pairs(values) do
          local id = normalize_id(k)
-         value[id] = struct[id]
+	 value[id] = struct[id]
```
extraneous tab

wingo commented on src/lib/yang/schema.lua:
```diff
@@ -540,6 +540,40 @@ declare_initializer(init_typedef, 'typedef')
 declare_initializer(init_value, 'value')
 
+
+-- 7.8.2.  The list's key Statement
```
Is the real point here to make sure that key leaves are mandatory?  I feel like this code's goals could be accomplished in less.  I also think it needs to run after the "resolve" phase so that any groupings are pulled in as appropriate.  Would you be OK if I PR'd to this one to shorten this up?

tsyesika replied:
Sure.

wingo commented on src/program/lwaftr/doc/configuration.md:
```diff
@@ -134,67 +142,42 @@ encapsulated in IPv6 and sent to the AFTR; the AFTR does the reverse.
-By default, a softwire is associated with the first `br-address`
-(`br 0;`).  To associate the tunnel with a different border router,
-specify it by index:
+   shift = `psid-length` + `reserved-ports-bit-count` - 16
```
Rather, 16 - psid-length - reserved-ports-bit-count

wingo commented on src/program/lwaftr/doc/configuration.md:
````diff
@@ -134,67 +142,42 @@ encapsulated in IPv6 and sent to the AFTR; the AFTR does the reverse.
-```
-  softwire { ipv4 178.79.150.233; psid 80; b4-ipv6 127:2:3:4:5:6:7:128; aftr 0; }
-```
+It's only necessary to specify `psid-length` as
````
psid-length also defaults to 16, no?  or at least it should, right?

wingo reviewed (commented):
Looks good.  A few nits, should we fix them up before merging?

Commit:
Fix some small nits

wingo committed:
yang: Mark list keys as mandatory after expansion

Delay the "sanitize_lists" pass to after children are resolved, and
inline it into the "resolve" pass.

wingo committed:
Merge pull request #818 from Igalia/mandatory-list-key

yang: Mark list keys as mandatory after expansion

wingo reviewed (approved):
my name is andy wingo and i approve this bodacious Pull Request
